scrapy/contrib/periodicstats/collectors.py

```diff
@@ -0,0 +1,157 @@
+                pipeline.open_spider(spider)
+            self._setup_looping_task(spider=spider, now=True)
+
+    def close_spider(self, spider, reason):
+        super(PeriodicStatsCollector, self).close_spider(spider, reason)
+        if self._enabled:
+            self._task.stop()
+            self._process_interval_stats(spider=spider, is_close=True)
+            for pipeline in self._pipelines:
+                pipeline.close_spider(spider, reason)
+
+    def set_value(self, key, value, spider=None):
+        super(PeriodicStatsCollector, self).set_value(key=key, value=value, spider=spider)
+        if self._enabled:
+            for observer in self._get_key_observers(key):
+                observer.set_value(value=value)
+
+    def inc_value(self, key, count=1, start=0, spider=None):
+        super(PeriodicStatsCollector, self).inc_value(key=key, count=count, start=start, spider=spider)
+        if self._enabled:
+            for observer in self._get_key_observers(key):
+                observer.inc_value(count=count, start=start)
+
+    def max_value(self, key, value, spider=None):
+        super(PeriodicStatsCollector, self).max_value(key=key, value=value, spider=spider)
+        if self._enabled:
+            self.set_value(key=key, value=self.get_value(key), spider=spider)
+
+    def min_value(self, key, value, spider=None):
+        super(PeriodicStatsCollector, self).min_value(key=key, value=value, spider=spider)
+        if self._enabled:
+            self.set_value(key=key, value=self.get_value(key), spider=spider)
+
+    def _setup_looping_task(self, _result=None, **kwargs):
+        spider = kwargs.pop('spider')
+        self._task = task.LoopingCall(self._process_interval_stats, spider)
+        d = self._task.start(1, **kwargs)
+        d.addErrback(self._setup_looping_task, spider=spider, now=False)
+
+    def _load_observers(self, crawler):
+        """
+        Loads the stats value observers
+        """
+        export_keys = set()
+        self._observers = defaultdict(list)
+        self._re_observers = defaultdict(list)
+        for observer in crawler.settings.get('PERIODIC_STATS_OBSERVERS', {}):
+            if not isinstance(observer, PeriodicStatsObserver):
+                raise NotConfigured('Stats observers must subclass PeriodicStatsObserver')
+            observer_export_key = observer.export_key
+            if observer_export_key in export_keys:
+                raise NotConfigured('Duplicate export key "%s"' % observer_export_key)
+            export_keys.add(observer_export_key)
+            if not observer.use_re_key:
+                self._observers[observer.key].append(observer)
+            else:
+                self._re_observers[observer.key].append(observer)
+
+    def _load_pipelines(self, crawler):
+        """
+        Loads the stats processor pipelines
+        """
+        self._pipelines = []
+        for pipeline in crawler.settings.get('PERIODIC_STATS_PIPELINES', []):
```

Comment:
[] should be defined as a default for PERIODIC_STATS_PIPELINES on scrapy/settings/default_settings.py

I think this should be a dictionary instead, mapping the pipeline and its priority in the execution order, to be consistent with ITEM_PIPELINES.